src/librustc/ty/inhabitedness.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::mem;
+use rustc_data_structures::small_vec::SmallVec;
+use syntax::ast::CRATE_NODE_ID;
+use util::nodemap::FxHashSet;
+use ty::context::TyCtxt;
+use ty::{AdtDef, VariantDef, FieldDef, TyS};
+use ty::{DefId, Substs};
+use ty::{AdtKind, Visibility, DefIdTree};
+use ty::TypeVariants::*;
+
+/// Represents a set of DefIds closed under the ancestor relation. That is, if
```

Comment:
Ah, I see the ancestor thing is based on the `DefPath` using `DefIdTree` (which -- preexisting, though -- prob ought to have some comments too...)